nimlsp/nimlsppkg/nimsuggest.nim

```diff
@@ -0,0 +1,299 @@
+    conf = newConfigRef(reportHook)
+    self = NimProg(
+      suggestMode: true,
+      processCmdLine: mockCmdLine
+    )
+  conf.writeHook = proc(conf: ConfigRef, s: string, flags: MsgFlags) = discard
+  conf.writelnHook = proc(conf: ConfigRef, s: string, flags: MsgFlags) = discard
+  conf.astDiagToLegacyReport = cli_reporter.legacyReportBridge
+  self.initDefinesProg(conf, "nimsuggest")
+
+  self.processCmdLineAndProjectPath(conf, [])
+
+  # Find Nim's prefix dir.
+  if nimPath == "":
+    let binaryPath = findExe("nim")
+    if binaryPath == "":
+      raise newException(IOError,
+          "Cannot find Nim standard library: Nim compiler not in PATH")
+    conf.prefixDir = AbsoluteDir binaryPath.splitPath().head.parentDir()
+    if not dirExists(conf.prefixDir / RelativeDir"lib"):
+      conf.prefixDir = AbsoluteDir""
+  else:
+    conf.prefixDir = AbsoluteDir nimPath
+
+  var graph = newModuleGraph(cache, conf)
+  graph.onMarkUsed = proc (g: ModuleGraph; info: TLineInfo; s: PSym; usageSym: var PSym; isDecl: bool) =
+    suggestSym(g, info, s, usageSym, isDecl)
+  graph.onSymImport = graph.onMarkUsed # same callback
+  if self.loadConfigsAndProcessCmdLine(cache, conf, graph, []):
+    customizeForBackend(graph, conf, backendC)
+    mockCommand(graph)
+
+  retval.doStopCompile = proc (): bool = false
+  return NimSuggest(graph: retval, idle: 0, cachedMsgs: cachedMsgs)
+
+proc getSymNode(node: ParsedNode): ParsedNode =
+  result = node
+  if result.kind == pnkPostfix:
+    result = result[^1]
+  elif result.kind == pnkPragmaExpr:
+    result = getSymNode(result[0])
+
+proc pnkToSymKind(kind: ParsedNodeKind): TSymKind =
+  result = skUnknown
+  case kind
+  of pnkConstSection, pnkConstDef: result = skConst
+  of pnkLetSection: result = skLet
+  of pnkVarSection: result = skVar
+  of pnkProcDef: result = skProc
+  of pnkFuncDef: result = skFunc
+  of pnkMethodDef: result = skMethod
+  of pnkConverterDef: result = skConverter
+  of pnkIteratorDef: result = skIterator
+  of pnkMacroDef: result = skMacro
+  of pnkTemplateDef: result = skTemplate
+  of pnkTypeDef, pnkTypeSection: result = skType
+  else: discard
+
+proc getName(node: ParsedNode): string =
+  if node.kind == pnkIdent:
+    result = node.startToken.ident.s
+  elif node.kind == pnkAccQuoted:
+    result = "`"
+    for t in node.idents:
+      result.add t.ident.s
+    result.add "`"
+
+proc processFlags(sug: Suggest; n: ParsedNode) =
+  var
+    identDeprecated: bool
+    colonDeprecated: bool
+  for s in n.sons:
+    identDeprecated = s.kind == pnkIdent and getName(s) == "deprecated"
+    colonDeprecated = s.kind == pnkExprColonExpr and getName(s[0]) == "deprecated" 
+    if identDeprecated or colonDeprecated:
+      sug.flags.incl SuggestFlag.deprecated
+
+proc parsedNodeToSugget(n: ParsedNode; originKind: ParsedNodeKind; module: string): Suggest =
+  if n.kind in {pnkError, pnkEmpty}: return
+  if n.kind notin {pnkConstSection..pnkTypeDef, pnkIdentDefs}: return
```

Comment:
use named sets, unnecessarily applying an ordering on the enum by one potential consumer makes for brittle code.